src/digraph.rs

```diff
@@ -1163,21 +1163,22 @@ impl PyDiGraph {
 
             let mut edges_to_add: Vec<(usize, usize, PyObject)> = Vec::new();
             for dir in &DIRECTIONS {
-                let edges;
-                if dir == &petgraph::Direction::Outgoing {
-                    edges = self.out_edges(u);
-                } else {
-                    edges = self.in_edges(u);
-                }
-
-                for edge in edges {
+                for edge in self.graph.edges_directed(NodeIndex::new(u), *dir) {
                     let s = edge.source();
                     let d = edge.target();
 
-                    if s == u {
-                        edges_to_add.push((v, d, edge.weight().clone()));
+                    if s.index() == u {
+                        edges_to_add.push((
+                            v,
+                            d.index(),
+                            edge.weight().clone_ref(py),
```

Comment:
Why `clone_ref`? Also, why does `index()` have to be added here but not before?

Reply:
This was working before because it was using `clone()` instead of `clone_ref()` Before this PR `in_edges()` and `out_edges()` were returning a `Vec(usize, usize, &PyObject)` and `edges_to_add` was a `Vec(usize, usize, &PyObject)` (it's just being used as a temp storage Vec before its used later to add edges). So prior to the changing `in_edges()` and `out_edges()` this worked by the 2 usizes implement copy so can efficiently be used in place without moving, but to store `&PyObject` as a `PyObject` (which is neededt for creating a new edge in the graph) it needs to take ownership of the object being borrowed. We can't do `*edge.weight()` since it would move the weight out of the graph (and would fail) so we need to clone/copy it which for PyObject isn't a big deal because it just increments the ref counter.

The change I made here (besides switching to petgraph method `edges_directed` instead of using in_edges and out_edges) is to switch just switch from `clone()` to `clone_ref(py)`. There isn't a functional difference between them as they both increment the python reference counter for the object, except that you give a gil handle to clone_ref if you have it: https://github.com/PyO3/pyo3/blob/v0.12.4/src/instance.rs#L196-L200 while `clone()` will queue it for when there is one it doesn't already have it: https://github.com/PyO3/pyo3/blob/v0.12.4/src/instance.rs#L517-L524. In the context of this method I don't think there is any performance difference since it has a gil reference, I just prefer it to be explicit and use clone_ref when we already have py

Reply:
Oh I missed the second half of the question. The `index()` calls are needed because the `edges_to_add` vec is defined as `Vec<(usize, usize, PyObject)>` and `self.graph.edges_directed()` returns an iterator yielding petgraph `EdgeReference` objects. Those objects return `NodeIndex` objects for `source()` and `target()` instead of usize ints. So to get the usize from the `NodeIndex` I had to call index() on it which returns a usize (since NodeIndex is just a wrapper around a usize anyay). This wasn't necessary before because in_edges() and out_edges returned a `Vec<usize, usize, &PyObject)>` and had already done the conversion from NodeIndex -> usize for us.

Reply:
got it!:)